{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PG-tGRnlFLA3"
   },
   "source": [
    "# Graident Boosting\n",
    "canonical lncRNA, -1 threshold, cross-valiation, all the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0RmwUsVLFLA6",
    "outputId": "052aeb8f-72c2-4689-d17e-a3dd86ec12ae"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-20 15:32:10.758751\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OlzN9OdsFWEU",
    "outputId": "f5b1ab2e-f40d-431f-c08e-4666ca7ffb34"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU device not found\n",
      "Running on CoLab\n",
      "Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount(\"/content/drive/\", force_remount=True).\n",
      "/content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "dt='float32'\n",
    "tf.keras.backend.set_floatx('float32')\n",
    "tf.random.set_seed(42) # supposedly leads to reproducible results\n",
    "\n",
    "device_name = tf.test.gpu_device_name()\n",
    "if device_name != '/device:GPU:0':\n",
    "    print('GPU device not found')\n",
    "else:\n",
    "    print('Found GPU at: {}'.format(device_name))\n",
    "\n",
    "try:\n",
    "    from google.colab import drive\n",
    "    IN_COLAB = True\n",
    "    print('Running on CoLab')\n",
    "    PATH='/content/drive/'\n",
    "    drive.mount(PATH)\n",
    "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
    "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
    "except:\n",
    "    IN_COLAB = False\n",
    "    DATA_DIR=\"/\"\n",
    "    MODEL_DIR=\"/\"\n",
    "print(DATA_DIR)\n",
    "SAVE_MODEL_FILENAME = None "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PRX-UEr8FLA8",
    "outputId": "341ba2c9-0dd6-49c2-ce55-c1661981c3b2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.9.16\n",
      "sklearn 1.2.2\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print('Python',python_version())\n",
    "import numpy as np\n",
    "np.random.seed(42) # supposedly sets scikit-learn\n",
    "import time # sleep function\n",
    "from os.path import isfile\n",
    "from matplotlib import pyplot as plt \n",
    "import sklearn   # pip install --upgrade scikit-learn\n",
    "print('sklearn',sklearn.__version__)\n",
    "\n",
    "from tensorflow import keras\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import auc\n",
    "from sklearn.metrics import matthews_corrcoef\n",
    "from sklearn.utils import class_weight\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.ensemble import GradientBoostingClassifier as GBC\n",
    "\n",
    "K=5\n",
    "ALPHABET=4**K + 1\n",
    "EPOCHS=150 \n",
    "RCI_THRESHOLD_MECHANISM = 'ZERO'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
    "RCI_THRESHOLD_VALUE = -1  # use -1 for lncRNA, use 0 for mRNA\n",
    "BREAK = False   # optionally break after first fold\n",
    "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
    "FILTER_TRAIN        = False\n",
    "FILTER_TAILS_TRAIN  = False\n",
    "FILTER_MIDDLE_TRAIN = False\n",
    "FILTER_TEST         = False\n",
    "FILTER_TAILS_TEST   = False\n",
    "FILTER_MIDDLE_TEST  = False\n",
    "MIDDLE_LOW          = -2   # -2 for lncRNA, -1 for mRNA\n",
    "MIDDLE_HIGH         = 0    #  0 for lncRNA, +1 for mRNA\n",
    "\n",
    "REPEATS = 2\n",
    "FOLDS = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "id": "LnkpVKdMFLA-"
   },
   "outputs": [],
   "source": [
    "RCI_FILE_TRAIN = 'train.lncRNA_RCI.all_cell_lines.csv'\n",
    "RCI_FILE_TEST  = None # use None for cross-validation\n",
    "\n",
    "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
    "SEQ_FILE_TEST  = None # use None for cross-validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "id": "e3p4QzQJFLA_"
   },
   "outputs": [],
   "source": [
    "def get_ordered_list():\n",
    "    ordered_list = \\\n",
    "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
    "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
    "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
    "    return ordered_list\n",
    "all_cell_lines = get_ordered_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UtqdpJOxFLBA"
   },
   "source": [
    "## Data Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "id": "p35ehKV3Kq0z"
   },
   "outputs": [],
   "source": [
    "class DataLoader():\n",
    "    def __init__(self):\n",
    "        self.cache=dict() \n",
    "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        self.gene2rci = dict()\n",
    "        \n",
    "    def load_gene_rci_values(self,filepath,exclusions):\n",
    "        '''\n",
    "        Load all the genes from the given RCI csv file.\n",
    "        The given file usually contains train or test, not both.\n",
    "        Return dict with keys=gene:str and values=RCI:float.\n",
    "        '''\n",
    "        self.gene2rci = {}\n",
    "        overall_sum = 0\n",
    "        with open (filepath,'r') as handle:\n",
    "            for row in handle:\n",
    "                line = row.strip()\n",
    "                fields = line.split(',')\n",
    "                gene_id = fields.pop(0)\n",
    "                cell_line_index = 0\n",
    "                rci_values = []\n",
    "                log_mean=0.0\n",
    "                for rci_str in fields:\n",
    "                    if cell_line_index not in exclusions:\n",
    "                        if rci_str != \"nan\":\n",
    "                            rci_val = float(rci_str)\n",
    "                            rci_values.append(rci_val)\n",
    "                    cell_line_index += 1\n",
    "                if len(rci_values)>0:\n",
    "                    values = np.array(rci_values)\n",
    "                    antilogs = [2**x for x in values]\n",
    "                    big_mean = np.mean(antilogs)\n",
    "                    # Avoid division by zero\n",
    "                    if np.absolute(big_mean)<0.000001:\n",
    "                        log_mean = -1000000 # neg infinity\n",
    "                    else:\n",
    "                        log_mean = np.log2(big_mean) \n",
    "                    self.gene2rci[gene_id] = log_mean\n",
    "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
    "        return self.gene2rci\n",
    "\n",
    "    def _seq_to_kmer_values(self,rna,K):\n",
    "        vec=[] # seq converted to list of K-mers \n",
    "        N_indicator = 0 # indicator value\n",
    "        length = len(rna)\n",
    "        for i in range(length-K+1):\n",
    "            kmer = rna[i:i+K]\n",
    "            if 'N' in kmer:\n",
    "                value = N_indicator\n",
    "            elif kmer in self.cache.keys():\n",
    "                value = self.cache[kmer]\n",
    "            else:\n",
    "                value = 0\n",
    "                for j in range(K):\n",
    "                    value *= 4   \n",
    "                    nextnuc = kmer[j] \n",
    "                    nucval = self.vals[nextnuc]\n",
    "                    value += nucval\n",
    "                value += 1   # NNN => 0, AAA => 1\n",
    "                self.cache[kmer] = value\n",
    "            vec.append(value)\n",
    "        return vec\n",
    "\n",
    "    def load_sequence(self,filepath):\n",
    "        '''\n",
    "        Load all the sequences from the given file. \n",
    "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
    "        Each line has IDs plus sequence.\n",
    "        The IDs in the file do not include dot-version numbers.\n",
    "        The file may already be filtered e.g. canonical transcripts.\n",
    "        '''\n",
    "        allids=[]\n",
    "        allseq=[]\n",
    "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row\n",
    "                else:\n",
    "                    line    = row.strip()\n",
    "                    fields  = line.split(',')\n",
    "                    tran_id = fields[0]  # without version number\n",
    "                    gene_id = fields[1]  # without version number\n",
    "                    seq_len = int(fields[3])\n",
    "                    seq_txt = fields[4]\n",
    "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
    "                    # We have validated this by spot checking.\n",
    "                    # TO DO: validate this programmatically.\n",
    "                    if gene_id in self.gene2rci.keys():\n",
    "                        allids.append( (gene_id,tran_id) )\n",
    "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
    "                        allseq.append(hot_vec)\n",
    "        self.cache=dict() # save RAM\n",
    "        return allids,allseq\n",
    "\n",
    "    def load_spectra(self,filepath):\n",
    "        '''\n",
    "        Load all (variable-length) sequences as lists of kmers.\n",
    "        Then convert each sequence to (fixed-length) kmer histograms.\n",
    "        '''\n",
    "        allids,allseq = self.load_sequence(filepath)\n",
    "        allspectra = []\n",
    "        for seq in allseq:\n",
    "            spectrum = np.zeros(ALPHABET)\n",
    "            for kmer in seq:\n",
    "                spectrum[kmer] += 1\n",
    "            spectrum /= len(seq)\n",
    "            allspectra.append(spectrum)\n",
    "        return allids,allspectra        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VDZ6siB_Kq04"
   },
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "id": "AwMbRjm0FLBF"
   },
   "outputs": [],
   "source": [
    "def build_model():\n",
    "    gbc = GBC()\n",
    "    return gbc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "clj-wufgFLBF",
    "outputId": "c4de939d-36d8-45a3-b656-c1edeea19897"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-20 15:32:12.339221\n",
      "GradientBoostingClassifier()\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "model=build_model()\n",
    "print(model)  # Print this only once\n",
    "model=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgrC1alOKq07"
   },
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "id": "W9xiFzNbFLBE"
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "class CrossValidator():\n",
    "    def __init__(self,epochs,score_threshold=0.5):\n",
    "        self.epochs = epochs\n",
    "        self.score_threshold = score_threshold\n",
    "        self.mechanism = 'ZERO'\n",
    "        self.discriminator = RCI_THRESHOLD_VALUE\n",
    "        self.flip = False\n",
    "        self.reset_statistics()\n",
    "        \n",
    "    def reset_statistics(self):\n",
    "        self.cv_accuracy=[]\n",
    "        self.cv_precision=[]\n",
    "        self.cv_recall=[]\n",
    "        self.cv_f1=[]\n",
    "        self.cv_auprc=[]\n",
    "        self.cv_auroc=[]\n",
    "        self.cv_mcc=[]\n",
    "        \n",
    "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
    "        # Prepare X and y for training or testing.\n",
    "        subsetX=[]\n",
    "        subsetY=[]\n",
    "        for t in range(len(all_ids)):\n",
    "            gene_id,tran_id = all_ids[t]\n",
    "            oneX            = all_seqs[t]\n",
    "            oneY            = rci_map[gene_id]\n",
    "            subsetX.append(oneX)\n",
    "            subsetY.append(oneY)\n",
    "        subsetX = np.array(subsetX)\n",
    "        subsetY = np.array(subsetY).reshape((-1,1))\n",
    "        return subsetX,subsetY\n",
    "    \n",
    "    def set_threshold_mechanism(self, mechanism):\n",
    "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
    "            raise Exception('Unrecognized mechansm:',mechanism)\n",
    "        self.mechanism = mechanism\n",
    "    \n",
    "    def _apply_threshold(self, array_of_rci):\n",
    "        # Takes list of float, returns list of labels [0,1].\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            labels = self.discriminator.predict(array_of_rci)\n",
    "            if self.flip:\n",
    "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
    "                labels = np.array(list(map(IS_CYTO, labels)))\n",
    "        else:  # 'THE_MEAN' or 'ZERO'\n",
    "            rci_threshold = self.discriminator\n",
    "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
    "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
    "        return labels\n",
    "    \n",
    "    def _prepare_threshold(self, rci_values, create=True):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            if create:  # during training, create a new GMM\n",
    "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
    "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
    "                gmm.fit(rci_values)\n",
    "            else:   # during testing, use existing GMM\n",
    "                gmm=self.discriminator\n",
    "            self.flip = False\n",
    "            # The GMM labels are arbitrary.\n",
    "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
    "                self.flip = True\n",
    "            self.discriminator = gmm   # redundant but consistent\n",
    "        elif self.mechanism == 'THE_MEAN':\n",
    "            self.discriminator = np.mean(rci_values)\n",
    "        elif self.mechanism == 'ZERO':\n",
    "            self.discriminator = RCI_THRESHOLD_VALUE\n",
    "        else: # not expected\n",
    "            self.discriminator = 0\n",
    "    \n",
    "    def _explain_threshold(self):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            gmm=self.discriminator\n",
    "            print('Discriminator is GMM')\n",
    "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
    "            print('Variances',gmm.covariances_)\n",
    "            print('Priors',gmm.weights_)\n",
    "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
    "            print(test_rcis)\n",
    "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
    "        else:\n",
    "            print('Discriminator',self.mechanism,self.discriminator)\n",
    "    \n",
    "    def _show_sizes(self,label,values):\n",
    "        a = np.count_nonzero(values==1)\n",
    "        b = np.count_nonzero(values==0)\n",
    "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
    "        \n",
    "    def save_model(self,filename):\n",
    "        if self.model is not None:\n",
    "            filepath = MODEL_DIR + filename\n",
    "            #self.model.save(filepath)\n",
    "            print('? Saved model to',filepath)\n",
    "        \n",
    "    def load_model(self,filename):\n",
    "        filepath = MODEL_DIR + filename\n",
    "        #self.model = keras.models.load_model(filepath)\n",
    "        print('? Loaded model from',filepath)\n",
    "        \n",
    "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
    "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
    "        print(datetime.now())\n",
    "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
    "        self._prepare_threshold(y_rci,True)  \n",
    "        self._explain_threshold()\n",
    "        y_train = self._apply_threshold(y_rci)\n",
    "        self._show_sizes('Train',y_train)\n",
    "        if valid_ids is not None:\n",
    "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
    "            y_valid = self._apply_threshold(y_rci)\n",
    "            self._show_sizes('Valid',y_valid)\n",
    "        y_rci = None\n",
    "\n",
    "        self.model=build_model()\n",
    "        \n",
    "        print(\"FIT\")\n",
    "        print(datetime.now())\n",
    "        self.model.fit(X_train, y_train) # sample weight\n",
    "\n",
    "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
    "        # For final test, do no train.\n",
    "        # Assume set_sequences() set the test set.\n",
    "        print(datetime.now())\n",
    "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
    "        y_test = self._apply_threshold(y_rci)\n",
    "        y_rci = None\n",
    "        \n",
    "        print(\"PREDICT\")\n",
    "        print(datetime.now())        \n",
    "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
    "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
    "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
    "        \n",
    "        print('debug pred',yhat_pred[:3])\n",
    "        print('debug class',yhat_classes[:3])\n",
    "\n",
    "        self._show_sizes('Test',y_test)\n",
    "        self._show_sizes('Predict',yhat_classes)\n",
    "        print('Test sizes',X_test.shape,y_test.shape)\n",
    "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
    "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
    "        print('Score threshold',self.score_threshold)\n",
    "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
    "        print('Confusion matrix\\n',cm1)\n",
    "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
    "        print('Normalized matrix\\n',cm2)\n",
    "\n",
    "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
    "        precision = precision_score(y_test, yhat_classes)*100.\n",
    "        recall = recall_score(y_test, yhat_classes)*100.\n",
    "        f1 = f1_score(y_test, yhat_classes)*100.\n",
    "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
    "        auprc = auc(prc_X,prc_Y)*100.\n",
    "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
    "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
    "\n",
    "        correct_pos = 0\n",
    "        correct_neg = 0\n",
    "        wrong_pos = 0\n",
    "        wrong_neg = 0\n",
    "        for i in range(len(y_test)):\n",
    "            if yhat_pred[i]>=0.65:\n",
    "                if y_test[i]==1:\n",
    "                    correct_pos += 1\n",
    "                else:\n",
    "                    wrong_pos += 1\n",
    "            elif yhat_pred[i]<=0.35:\n",
    "                if y_test[i]==0:\n",
    "                    correct_neg += 1\n",
    "                else:\n",
    "                    wrong_neg += 1\n",
    "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
    "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
    "\n",
    "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
    "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
    "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
    "\n",
    "        self.cv_accuracy.append(accuracy)\n",
    "        self.cv_precision.append(precision)\n",
    "        self.cv_recall.append(recall)\n",
    "        self.cv_f1.append(f1)\n",
    "        self.cv_mcc.append(mcc)\n",
    "        self.cv_auprc.append(auprc)\n",
    "        self.cv_auroc.append(auroc)\n",
    "\n",
    "    def get_statistics(self):\n",
    "        return \\\n",
    "        self.cv_accuracy,\\\n",
    "        self.cv_precision,\\\n",
    "        self.cv_recall,\\\n",
    "        self.cv_f1,\\\n",
    "        self.cv_mcc,\\\n",
    "        self.cv_auprc,\\\n",
    "        self.cv_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Td97uyyj5qDq"
   },
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "id": "poe3rwAF4Ir7"
   },
   "outputs": [],
   "source": [
    "class Separator():\n",
    "    def __init__(self):\n",
    "        self.train_ids = []\n",
    "        self.train_seq = []\n",
    "        self.train_rci = dict()\n",
    "        self.val_ids = []\n",
    "        self.val_seq = []\n",
    "        self.val_rci = dict()\n",
    "    def load(self,data_dir,rep,fold):\n",
    "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
    "        filename = data_dir + filename\n",
    "        self.val_genes = set()\n",
    "        print('Opening file',filename)\n",
    "        with open(filename,'r') as fin:\n",
    "            for line in fin:\n",
    "                gene_id = line.strip()\n",
    "                self.val_genes.add(gene_id)\n",
    "    def process(self,allids,allseq,gene_to_rci):\n",
    "        size = len(allids)\n",
    "        for t in range(size):\n",
    "            gene_id,tran_id = allids[t]\n",
    "            oneX            = allseq[t]\n",
    "            oneY            = gene_to_rci[gene_id]\n",
    "            in_middle = gene_to_rci[gene_id] >= MIDDLE_LOW and gene_to_rci[gene_id] <= MIDDLE_HIGH\n",
    "            in_tails = gene_to_rci[gene_id] < MIDDLE_LOW or gene_to_rci[gene_id] > MIDDLE_HIGH\n",
    "            if gene_id in self.val_genes:\n",
    "                if FILTER_TEST and (\\\n",
    "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.val_ids.append(allids[t])\n",
    "                    self.val_seq.append(allseq[t])\n",
    "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
    "            else:\n",
    "                if FILTER_TRAIN and (\\\n",
    "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.train_ids.append(allids[t])\n",
    "                    self.train_seq.append(allseq[t])\n",
    "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
    "    def get_ids(self):\n",
    "        return self.train_ids,self.val_ids\n",
    "    def get_seq(self):\n",
    "        return self.train_seq,self.val_seq\n",
    "    def get_rci(self):\n",
    "        return self.train_rci,self.val_rci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XC9m0W-pFLBH",
    "outputId": "e27842f4-74de-41be-b0d4-33509adbf4f9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-20 15:32:12.464741\n",
      "Load RCI from /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.all_cell_lines.csv\n",
      "Number of RCI values loaded 4372\n",
      "Num RCI: 4372\n",
      "Load sequence from /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
      "Loaded 4372 sequences.\n",
      "\n",
      "Training # 1 1\n",
      "2023-04-20 15:32:20.207001\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
      "2023-04-20 15:32:20.220750\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1810:1687 51.76%\n",
      "Valid 1:0 433:442 49.49%\n",
      "FIT\n",
      "2023-04-20 15:32:20.249517\n",
      "\n",
      "Testing # 1 1\n",
      "2023-04-20 15:33:55.852421\n",
      "2023-04-20 15:33:55.852483\n",
      "PREDICT\n",
      "2023-04-20 15:33:55.858702\n",
      "debug pred [0.5694841403464626, 0.5741203360428534, 0.42563323866808467]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 433:442 49.49%\n",
      "Predict 1:0 497:378 56.80%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.5251482423504985 mean 0.1488576129974603 std\n",
      "Range of scores: 0.11676588201815573 to 0.8880124542681984\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[236 206]\n",
      " [142 291]]\n",
      "Normalized matrix\n",
      " [[0.26971429 0.23542857]\n",
      " [0.16228571 0.33257143]]\n",
      "Extreme scores correct, pos:neg 124 86\n",
      "Extreme scores incorrect pos:neg 58 32\n",
      "Accuracy: 60.23% Precision: 58.55% Recall: 67.21%\n",
      "F1: 62.58% MCC: 0.2079\n",
      "AUPRC: 62.28% AUROC: 65.02%\n",
      " accuracy [60.22857142857143]\n",
      " precision [58.55130784708249]\n",
      " recall [67.2055427251732]\n",
      " F1 [62.58064516129031]\n",
      " MCC [0.20791293492439555]\n",
      " AUPRC [62.280401274887275]\n",
      " AUROC [65.01781739521178]\n",
      "\n",
      "Training # 1 2\n",
      "2023-04-20 15:33:55.891976\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
      "2023-04-20 15:33:55.911863\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1789:1708 51.16%\n",
      "Valid 1:0 454:421 51.89%\n",
      "FIT\n",
      "2023-04-20 15:33:55.938972\n",
      "\n",
      "Testing # 1 2\n",
      "2023-04-20 15:35:30.273887\n",
      "2023-04-20 15:35:30.273951\n",
      "PREDICT\n",
      "2023-04-20 15:35:30.286829\n",
      "debug pred [0.5097987892989085, 0.47457406910386746, 0.7152842494536374]\n",
      "debug class [1 0 1]\n",
      "Test 1:0 454:421 51.89%\n",
      "Predict 1:0 439:436 50.17%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.5035047352373142 mean 0.1491187716589897 std\n",
      "Range of scores: 0.11160691132710561 to 0.8720017288431725\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[257 164]\n",
      " [179 275]]\n",
      "Normalized matrix\n",
      " [[0.29371429 0.18742857]\n",
      " [0.20457143 0.31428571]]\n",
      "Extreme scores correct, pos:neg 110 96\n",
      "Extreme scores incorrect pos:neg 44 39\n",
      "Accuracy: 60.80% Precision: 62.64% Recall: 60.57%\n",
      "F1: 61.59% MCC: 0.2160\n",
      "AUPRC: 65.33% AUROC: 64.70%\n",
      " accuracy [60.8]\n",
      " precision [62.64236902050114]\n",
      " recall [60.57268722466961]\n",
      " F1 [61.590145576707734]\n",
      " MCC [0.2160256512658388]\n",
      " AUPRC [65.32873293945603]\n",
      " AUROC [64.70434354955162]\n",
      "\n",
      "Training # 1 3\n",
      "2023-04-20 15:35:30.342443\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
      "2023-04-20 15:35:30.372588\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1810:1688 51.74%\n",
      "Valid 1:0 433:441 49.54%\n",
      "FIT\n",
      "2023-04-20 15:35:30.419772\n",
      "\n",
      "Testing # 1 3\n",
      "2023-04-20 15:37:05.492921\n",
      "2023-04-20 15:37:05.492980\n",
      "PREDICT\n",
      "2023-04-20 15:37:05.504314\n",
      "debug pred [0.5240928680737212, 0.5361826223831656, 0.2427711968521071]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 433:441 49.54%\n",
      "Predict 1:0 477:397 54.58%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5167856335604674 mean 0.14502777827309044 std\n",
      "Range of scores: 0.1457798243632182 to 0.9180937161858646\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[249 192]\n",
      " [148 285]]\n",
      "Normalized matrix\n",
      " [[0.28489703 0.21967963]\n",
      " [0.16933638 0.32608696]]\n",
      "Extreme scores correct, pos:neg 109 80\n",
      "Extreme scores incorrect pos:neg 61 33\n",
      "Accuracy: 61.10% Precision: 59.75% Recall: 65.82%\n",
      "F1: 62.64% MCC: 0.2238\n",
      "AUPRC: 61.82% AUROC: 64.61%\n",
      " accuracy [61.098398169336384]\n",
      " precision [59.74842767295597]\n",
      " recall [65.81986143187068]\n",
      " F1 [62.637362637362635]\n",
      " MCC [0.22375444512373463]\n",
      " AUPRC [61.81678836903096]\n",
      " AUROC [64.61328180232833]\n",
      "\n",
      "Training # 1 4\n",
      "2023-04-20 15:37:05.542699\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
      "2023-04-20 15:37:05.561518\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1778:1720 50.83%\n",
      "Valid 1:0 465:409 53.20%\n",
      "FIT\n",
      "2023-04-20 15:37:05.586439\n",
      "\n",
      "Testing # 1 4\n",
      "2023-04-20 15:38:41.954055\n",
      "2023-04-20 15:38:41.954117\n",
      "PREDICT\n",
      "2023-04-20 15:38:41.960473\n",
      "debug pred [0.5092138372462951, 0.25982360996551135, 0.8094368219340654]\n",
      "debug class [1 0 1]\n",
      "Test 1:0 465:409 53.20%\n",
      "Predict 1:0 455:419 52.06%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5073338299038873 mean 0.1462720109666054 std\n",
      "Range of scores: 0.14608385078504618 to 0.8651968057244009\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[238 171]\n",
      " [181 284]]\n",
      "Normalized matrix\n",
      " [[0.27231121 0.19565217]\n",
      " [0.20709382 0.32494279]]\n",
      "Extreme scores correct, pos:neg 103 88\n",
      "Extreme scores incorrect pos:neg 42 45\n",
      "Accuracy: 59.73% Precision: 62.42% Recall: 61.08%\n",
      "F1: 61.74% MCC: 0.1924\n",
      "AUPRC: 65.11% AUROC: 64.08%\n",
      " accuracy [59.7254004576659]\n",
      " precision [62.417582417582416]\n",
      " recall [61.0752688172043]\n",
      " F1 [61.73913043478261]\n",
      " MCC [0.1924272075704105]\n",
      " AUPRC [65.11207732654233]\n",
      " AUROC [64.08391829008596]\n",
      "\n",
      "Training # 1 5\n",
      "2023-04-20 15:38:41.998958\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
      "2023-04-20 15:38:42.013196\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1785:1713 51.03%\n",
      "Valid 1:0 458:416 52.40%\n",
      "FIT\n",
      "2023-04-20 15:38:42.038336\n",
      "\n",
      "Testing # 1 5\n",
      "2023-04-20 15:40:17.019678\n",
      "2023-04-20 15:40:17.020113\n",
      "PREDICT\n",
      "2023-04-20 15:40:17.028498\n",
      "debug pred [0.233853457494427, 0.6202752650369819, 0.3199728068233959]\n",
      "debug class [0 1 0]\n",
      "Test 1:0 458:416 52.40%\n",
      "Predict 1:0 463:411 52.97%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5120590353975222 mean 0.13997060455464394 std\n",
      "Range of scores: 0.12018123686151017 to 0.8775945126729084\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[244 172]\n",
      " [167 291]]\n",
      "Normalized matrix\n",
      " [[0.2791762  0.19679634]\n",
      " [0.19107551 0.33295195]]\n",
      "Extreme scores correct, pos:neg 105 72\n",
      "Extreme scores incorrect pos:neg 48 35\n",
      "Accuracy: 61.21% Precision: 62.85% Recall: 63.54%\n",
      "F1: 63.19% MCC: 0.2220\n",
      "AUPRC: 64.70% AUROC: 64.52%\n",
      " accuracy [61.212814645308924]\n",
      " precision [62.850971922246224]\n",
      " recall [63.53711790393013]\n",
      " F1 [63.192182410423456]\n",
      " MCC [0.22204662059108513]\n",
      " AUPRC [64.70346830619698]\n",
      " AUROC [64.51650151158884]\n",
      "\n",
      "Training # 2 1\n",
      "2023-04-20 15:40:17.075862\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
      "2023-04-20 15:40:17.100483\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1807:1690 51.67%\n",
      "Valid 1:0 436:439 49.83%\n",
      "FIT\n",
      "2023-04-20 15:40:17.138793\n",
      "\n",
      "Testing # 2 1\n",
      "2023-04-20 15:41:51.904983\n",
      "2023-04-20 15:41:51.905039\n",
      "PREDICT\n",
      "2023-04-20 15:41:51.911775\n",
      "debug pred [0.8669365795404583, 0.3946375639807145, 0.1647903210157286]\n",
      "debug class [1 0 0]\n",
      "Test 1:0 436:439 49.83%\n",
      "Predict 1:0 464:411 53.03%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.5120153759296462 mean 0.1443524195700817 std\n",
      "Range of scores: 0.0994647318735998 to 0.8756970733473703\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[246 193]\n",
      " [165 271]]\n",
      "Normalized matrix\n",
      " [[0.28114286 0.22057143]\n",
      " [0.18857143 0.30971429]]\n",
      "Extreme scores correct, pos:neg 103 93\n",
      "Extreme scores incorrect pos:neg 50 40\n",
      "Accuracy: 59.09% Precision: 58.41% Recall: 62.16%\n",
      "F1: 60.22% MCC: 0.1823\n",
      "AUPRC: 61.26% AUROC: 63.70%\n",
      " accuracy [59.08571428571429]\n",
      " precision [58.4051724137931]\n",
      " recall [62.15596330275229]\n",
      " F1 [60.222222222222236]\n",
      " MCC [0.18225767809120075]\n",
      " AUPRC [61.26309498847715]\n",
      " AUROC [63.69825082025453]\n",
      "\n",
      "Training # 2 2\n",
      "2023-04-20 15:41:51.942599\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
      "2023-04-20 15:41:51.963960\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1794:1703 51.30%\n",
      "Valid 1:0 449:426 51.31%\n",
      "FIT\n",
      "2023-04-20 15:41:51.989581\n",
      "\n",
      "Testing # 2 2\n",
      "2023-04-20 15:43:27.169905\n",
      "2023-04-20 15:43:27.169970\n",
      "PREDICT\n",
      "2023-04-20 15:43:27.176007\n",
      "debug pred [0.6699636951868508, 0.4712821992762059, 0.2303060300819319]\n",
      "debug class [1 0 0]\n",
      "Test 1:0 449:426 51.31%\n",
      "Predict 1:0 498:377 56.91%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.5275511623824267 mean 0.148499559690215 std\n",
      "Range of scores: 0.12678333267270855 to 0.8784141098229319\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[240 186]\n",
      " [137 312]]\n",
      "Normalized matrix\n",
      " [[0.27428571 0.21257143]\n",
      " [0.15657143 0.35657143]]\n",
      "Extreme scores correct, pos:neg 137 82\n",
      "Extreme scores incorrect pos:neg 63 23\n",
      "Accuracy: 63.09% Precision: 62.65% Recall: 69.49%\n",
      "F1: 65.89% MCC: 0.2607\n",
      "AUPRC: 67.28% AUROC: 67.02%\n",
      " accuracy [63.08571428571429]\n",
      " precision [62.65060240963856]\n",
      " recall [69.48775055679288]\n",
      " F1 [65.89229144667371]\n",
      " MCC [0.26067299948106265]\n",
      " AUPRC [67.27981264795146]\n",
      " AUROC [67.01642669678054]\n",
      "\n",
      "Training # 2 3\n",
      "2023-04-20 15:43:27.210348\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
      "2023-04-20 15:43:27.227573\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1771:1727 50.63%\n",
      "Valid 1:0 472:402 54.00%\n",
      "FIT\n",
      "2023-04-20 15:43:27.253203\n",
      "\n",
      "Testing # 2 3\n",
      "2023-04-20 15:45:02.533449\n",
      "2023-04-20 15:45:02.533874\n",
      "PREDICT\n",
      "2023-04-20 15:45:02.542406\n",
      "debug pred [0.3867737499430496, 0.506093224353497, 0.7028057585412342]\n",
      "debug class [0 1 1]\n",
      "Test 1:0 472:402 54.00%\n",
      "Predict 1:0 436:438 49.89%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.49834751535908883 mean 0.1499794637822793 std\n",
      "Range of scores: 0.08641496008387843 to 0.8887941128302826\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[242 160]\n",
      " [196 276]]\n",
      "Normalized matrix\n",
      " [[0.27688787 0.18306636]\n",
      " [0.22425629 0.31578947]]\n",
      "Extreme scores correct, pos:neg 103 95\n",
      "Extreme scores incorrect pos:neg 45 50\n",
      "Accuracy: 59.27% Precision: 63.30% Recall: 58.47%\n",
      "F1: 60.79% MCC: 0.1861\n",
      "AUPRC: 66.01% AUROC: 63.20%\n",
      " accuracy [59.26773455377574]\n",
      " precision [63.30275229357798]\n",
      " recall [58.47457627118644]\n",
      " F1 [60.79295154185023]\n",
      " MCC [0.18613641346345336]\n",
      " AUPRC [66.01222711530176]\n",
      " AUROC [63.204633611603]\n",
      "\n",
      "Training # 2 4\n",
      "2023-04-20 15:45:02.589894\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
      "2023-04-20 15:45:02.617939\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1816:1682 51.92%\n",
      "Valid 1:0 427:447 48.86%\n",
      "FIT\n",
      "2023-04-20 15:45:02.656079\n",
      "\n",
      "Testing # 2 4\n",
      "2023-04-20 15:46:37.852131\n",
      "2023-04-20 15:46:37.852196\n",
      "PREDICT\n",
      "2023-04-20 15:46:37.858430\n",
      "debug pred [0.44818266802196893, 0.46438860747443655, 0.7920324333566064]\n",
      "debug class [0 0 1]\n",
      "Test 1:0 427:447 48.86%\n",
      "Predict 1:0 440:434 50.34%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5071641746603193 mean 0.1538581090810385 std\n",
      "Range of scores: 0.1418164087120602 to 0.8805716755434828\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[262 185]\n",
      " [172 255]]\n",
      "Normalized matrix\n",
      " [[0.29977117 0.21167048]\n",
      " [0.19679634 0.29176201]]\n",
      "Extreme scores correct, pos:neg 110 103\n",
      "Extreme scores incorrect pos:neg 58 47\n",
      "Accuracy: 59.15% Precision: 57.95% Recall: 59.72%\n",
      "F1: 58.82% MCC: 0.1833\n",
      "AUPRC: 59.42% AUROC: 62.57%\n",
      " accuracy [59.153318077803206]\n",
      " precision [57.95454545454546]\n",
      " recall [59.71896955503513]\n",
      " F1 [58.82352941176471]\n",
      " MCC [0.183275764772756]\n",
      " AUPRC [59.42319951446768]\n",
      " AUROC [62.57118756843699]\n",
      "\n",
      "Training # 2 5\n",
      "2023-04-20 15:46:37.893470\n",
      "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
      "2023-04-20 15:46:37.913164\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1784:1714 51.00%\n",
      "Valid 1:0 459:415 52.52%\n",
      "FIT\n",
      "2023-04-20 15:46:37.937998\n",
      "\n",
      "Testing # 2 5\n",
      "2023-04-20 15:48:13.369431\n",
      "2023-04-20 15:48:13.369490\n",
      "PREDICT\n",
      "2023-04-20 15:48:13.377364\n",
      "debug pred [0.6113663590576492, 0.580601367183373, 0.5094337963608888]\n",
      "debug class [1 1 1]\n",
      "Test 1:0 459:415 52.52%\n",
      "Predict 1:0 492:382 56.29%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5204159294738929 mean 0.1404261533027947 std\n",
      "Range of scores: 0.10913536950623877 to 0.8852684913072777\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[232 183]\n",
      " [150 309]]\n",
      "Normalized matrix\n",
      " [[0.26544622 0.20938215]\n",
      " [0.17162471 0.35354691]]\n",
      "Extreme scores correct, pos:neg 121 73\n",
      "Extreme scores incorrect pos:neg 49 33\n",
      "Accuracy: 61.90% Precision: 62.80% Recall: 67.32%\n",
      "F1: 64.98% MCC: 0.2338\n",
      "AUPRC: 66.63% AUROC: 65.19%\n",
      " accuracy [61.89931350114416]\n",
      " precision [62.80487804878049]\n",
      " recall [67.3202614379085]\n",
      " F1 [64.98422712933754]\n",
      " MCC [0.23380342039435967]\n",
      " AUPRC [66.63185649103126]\n",
      " AUROC [65.1925348452634]\n",
      "2023-04-20 15:48:13.412165\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "\n",
    "accuracy=[]\n",
    "precision=[]\n",
    "recall=[]\n",
    "f1=[]\n",
    "mcc=[]\n",
    "auprc=[]\n",
    "auroc=[]\n",
    "\n",
    "loader = DataLoader()\n",
    "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
    "print(\"Load RCI from\",filepath)\n",
    "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "print('Num RCI:', len(gene_to_rci.keys()))\n",
    "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
    "print('Load sequence from',filepath)\n",
    "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
    "print('Loaded',len(allseq),'sequences.')\n",
    "test_gene_to_rci = None\n",
    "test_allids = None\n",
    "test_allseq = None\n",
    "if SEQ_FILE_TEST is not None:\n",
    "    # Train on the entire train set (no cross-validation).\n",
    "    # Evaluate with the test files.\n",
    "    test_loader = DataLoader()\n",
    "    filepath = DATA_DIR+RCI_FILE_TEST\n",
    "    print(\"Load RCI from\",filepath)\n",
    "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
    "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
    "    print('Load sequence from',filepath)\n",
    "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
    "    print('Loaded',len(test_allseq),'sequences.')\n",
    "\n",
    "for repeat in range(REPEATS):\n",
    "    for fold in range(FOLDS):\n",
    "        show_r = repeat+1  # display one-based counting\n",
    "        show_f = fold+1    # display one-based counting\n",
    "\n",
    "        print()\n",
    "        print(\"Training #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo = CrossValidator(EPOCHS)\n",
    "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
    "        if SEQ_FILE_TEST is None:\n",
    "            # Train on 80% and evaluate on 20%.\n",
    "            separator = Separator()\n",
    "            separator.load(DATA_DIR,show_r,show_f)\n",
    "            separator.process(allids,allseq,gene_to_rci)\n",
    "            train_allids,test_allids = separator.get_ids()\n",
    "            train_allseq,test_allseq = separator.get_seq()\n",
    "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                test_allids,test_allseq,test_gene_to_rci)\n",
    "            if SAVE_MODEL_FILENAME is not None:\n",
    "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
    "                cvdo.save_model(filename)\n",
    "        else:\n",
    "            # Train on the entire train set (no cross-validation).\n",
    "            # Evaluate with the test files.\n",
    "            train_allids = allids\n",
    "            train_allseq = allseq\n",
    "            train_gene_to_rci = gene_to_rci\n",
    "            BREAK = True\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                None,None,None)\n",
    "\n",
    "        print()\n",
    "        print(\"Testing #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo.reset_statistics()\n",
    "        cvdo.test_without_training(\n",
    "            test_allids,test_allseq,test_gene_to_rci)\n",
    "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
    "            cvdo.get_statistics()\n",
    "\n",
    "        print(\" accuracy\" ,  cv_accuracy)\n",
    "        print(\" precision\" , cv_precision)\n",
    "        print(\" recall\" ,    cv_recall)\n",
    "        print(\" F1\" ,        cv_f1)\n",
    "        print(\" MCC\" ,       cv_mcc)\n",
    "        print(\" AUPRC\" ,     cv_auprc)\n",
    "        print(\" AUROC\" ,     cv_auroc)\n",
    "\n",
    "        accuracy.append(cv_accuracy)\n",
    "        precision.append(cv_precision)\n",
    "        recall.append(cv_recall)\n",
    "        f1.append(cv_f1)\n",
    "        mcc.append(cv_mcc)\n",
    "        auprc.append(cv_auprc)\n",
    "        auroc.append(cv_auroc)\n",
    "        if BREAK: break\n",
    "    if BREAK: break\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HkCeDg_HdQ36",
    "outputId": "5cc0d4ab-6b86-45a4-af16-32f058b50fe3"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " accuracy mean 60.56% std 1.32\n",
      " precision mean 61.13% std 2.18\n",
      " recall mean 63.54% std 3.73\n",
      " F1 mean 62.25% std 2.12\n",
      " MCC mean 0.2108 std 0.0255\n",
      " AUPRC mean 63.99% std 2.61\n",
      " AUROC mean 64.46% std 1.22\n",
      " accuracy [[60.22857142857143], [60.8], [61.098398169336384], [59.7254004576659], [61.212814645308924], [59.08571428571429], [63.08571428571429], [59.26773455377574], [59.153318077803206], [61.89931350114416]]\n",
      " precision [[58.55130784708249], [62.64236902050114], [59.74842767295597], [62.417582417582416], [62.850971922246224], [58.4051724137931], [62.65060240963856], [63.30275229357798], [57.95454545454546], [62.80487804878049]]\n",
      " recall [[67.2055427251732], [60.57268722466961], [65.81986143187068], [61.0752688172043], [63.53711790393013], [62.15596330275229], [69.48775055679288], [58.47457627118644], [59.71896955503513], [67.3202614379085]]\n",
      " F1 [[62.58064516129031], [61.590145576707734], [62.637362637362635], [61.73913043478261], [63.192182410423456], [60.222222222222236], [65.89229144667371], [60.79295154185023], [58.82352941176471], [64.98422712933754]]\n",
      " MCC [[0.20791293492439555], [0.2160256512658388], [0.22375444512373463], [0.1924272075704105], [0.22204662059108513], [0.18225767809120075], [0.26067299948106265], [0.18613641346345336], [0.183275764772756], [0.23380342039435967]]\n",
      " AUPRC [[62.280401274887275], [65.32873293945603], [61.81678836903096], [65.11207732654233], [64.70346830619698], [61.26309498847715], [67.27981264795146], [66.01222711530176], [59.42319951446768], [66.63185649103126]]\n",
      " AUROC [[65.01781739521178], [64.70434354955162], [64.61328180232833], [64.08391829008596], [64.51650151158884], [63.69825082025453], [67.01642669678054], [63.204633611603], [62.57118756843699], [65.1925348452634]]\n"
     ]
    }
   ],
   "source": [
    "def STD (values):\n",
    "    # ddof=1 reduces bias when extrapolating from sample to population\n",
    "    return np.std(values,ddof=1)\n",
    "\n",
    "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
    "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
    "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
    "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
    "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
    "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
    "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
    "\n",
    "print(\" accuracy\"  , accuracy)\n",
    "print(\" precision\" , precision)\n",
    "print(\" recall\"    , recall)\n",
    "print(\" F1\"        , f1)\n",
    "print(\" MCC\"       , mcc)\n",
    "print(\" AUPRC\"     , auprc)\n",
    "print(\" AUROC\"     , auroc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "id": "QjSVa72v4IsA"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
